pylsp_jsonrpc/endpoint.py

```diff
@@ -14,6 +16,34 @@
 CANCEL_METHOD = '$/cancelRequest'
 
 
+async def run_as_daemon(func, *args):
+    future = futures.Future()
+    future.set_running_or_notify_cancel()
+
+    # A bug in python 3.7 makes it a bad idea to set a BaseException
+    # in a wrapped future (see except statement in asyncio.Task.__wakeup)
+    # Instead, we'll wrap base exceptions into exceptions and unwrap them
+    # on the other side of the call.
+    class BaseExceptionWrapper(Exception):
+        pass
+
+    def daemon():
+        try:
+            result = func(*args)
+        except Exception as e:
+            future.set_exception(e)
+        except BaseException as e:
+            future.set_exception(BaseExceptionWrapper(e))
+        else:
+            future.set_result(result)
+
+    Thread(target=daemon, daemon=True).start()
+    try:
+        return await asyncio.wrap_future(future)
```

Comment:
Also, why not use a Task for this? https://docs.python.org/3/library/asyncio-task.html#asyncio.create_task

Reply:
If a CPU-bound task is being run here, it is better to use `asyncio.loop.run_in_executor` https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.run_in_executor

Reply:
run_in_executor seems to work here for all use cases.

> Couldn't this future be awaited elsewhere and just return it as opposed to awaiting it?

Since I'm not quite sure every message handler only uses thread save data I just wanted to implement a way to retrieve the messages asynchronously, but still handle them sequentially.